File: Source/BleMidiGoodies/Private/Android/Java/BleMidiManager.java
package com.ninevastudios.blemidilib;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.bluetooth.BluetoothAdapter;
import android.content.pm.PackageManager;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import jp.kshoji.blemidi.central.BleMidiCentralProvider;
import jp.kshoji.blemidi.device.MidiInputDevice;
import jp.kshoji.blemidi.device.MidiOutputDevice;
import jp.kshoji.blemidi.listener.OnMidiDeviceAttachedListener;
import jp.kshoji.blemidi.listener.OnMidiDeviceDetachedListener;
import jp.kshoji.blemidi.listener.OnMidiScanStatusListener;

@Keep
public class BleMidiManager {

	public static native void OnInputDeviceEvent(InputDevice device);
	public static native void OnOutputDeviceEvent(OutputDevice device);
	public static native void OnMidiScanStatusChanged(boolean status);
	private BleMidiCentralProvider bleMidiCentralProvider;
	private final BluetoothAdapter mBluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
	private OnMidiScanStatusListener scanStatusListener;
	private OnMidiDeviceAttachedListener deviceAttachedListener;
	private OnMidiDeviceDetachedListener deviceDetachedListener;
	private boolean isInitialized = false;
	public BleMidiManager() {
	}

	@Keep
	public boolean initialize(Activity context) {
		if (!isBleSupported(context) || !isBluetoothEnabled())
			return false;

		bleMidiCentralProvider = new BleMidiCentralProvider(context);
		isInitialized = true;

		scanStatusListener = new OnMidiScanStatusListener() {
			@Override
			public void onMidiScanStatusChanged(boolean status) {
				OnMidiScanStatusChanged(status);
			}
		};

		bleMidiCentralProvider.setOnMidiScanStatusListener(scanStatusListener);

		deviceAttachedListener = new OnMidiDeviceAttachedListener() {
			@Override
			public void onMidiInputDeviceAttached(@NonNull MidiInputDevice midiInputDevice) {
				OnInputDeviceEvent(new InputDevice(midiInputDevice));
			}

			@Override
			public void onMidiOutputDeviceAttached(@NonNull MidiOutputDevice midiOutputDevice) {
				OnOutputDeviceEvent(new OutputDevice(midiOutputDevice));
			}
		};

		bleMidiCentralProvider.setOnMidiDeviceAttachedListener(deviceAttachedListener);

		deviceDetachedListener = new OnMidiDeviceDetachedListener() {
			@Override
			public void onMidiInputDeviceDetached(@NonNull MidiInputDevice midiInputDevice) {
				OnInputDeviceEvent(new InputDevice(midiInputDevice));
			}

			@Override
			public void onMidiOutputDeviceDetached(@NonNull MidiOutputDevice midiOutputDevice) {
				OnOutputDeviceEvent(new OutputDevice(midiOutputDevice));
			}
		};

		bleMidiCentralProvider.setOnMidiDeviceDetachedListener(deviceDetachedListener);

		return true;
	}

	@Keep
	public boolean isBleSupported(Activity activity) {
		return activity.getPackageManager().hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE);
	}

	@Keep
	public boolean isBluetoothEnabled() {
		return mBluetoothAdapter != null && mBluetoothAdapter.isEnabled();
	}

	@Keep
	@SuppressLint("MissingPermission")
	public void setBluetoothState(boolean isEnabled) {
		if (mBluetoothAdapter == null)
			return;

		if (isEnabled) {
			mBluetoothAdapter.enable();
		} else {
			mBluetoothAdapter.disable();
		}
	}

	@Keep
	public void setRequestPairing(boolean needsPairing) {
		if (!isInitialized)
			return;

		bleMidiCentralProvider.setRequestPairing(needsPairing);
	}

	@Keep
	public void startScan(int timeoutInMilliSeconds) {
		if (!isInitialized)
			return;

		bleMidiCentralProvider.startScanDevice(timeoutInMilliSeconds);
	}

	@Keep
	public void stopScan() {
		if (!isInitialized)
			return;

		bleMidiCentralProvider.stopScanDevice();
	}

	@Keep
	public void disconnectInputDevice(InputDevice midiInputDevice) {
		if (!isInitialized)
			return;

		bleMidiCentralProvider.disconnectDevice(midiInputDevice.getDevice());
	}

	@Keep
	public void disconnectOutputDevice(OutputDevice midiOutputDevice) {
		if (!isInitialized)
			return;

		bleMidiCentralProvider.disconnectDevice(midiOutputDevice.getDevice());
	}

	@Keep
	public InputDevice[] getMidiInputDevices() {
		ArrayList<InputDevice> result = new ArrayList<>();

		if (!isInitialized)
			return result.toArray(new InputDevice[0]);

		for (MidiInputDevice device : bleMidiCentralProvider.getMidiInputDevices()) {
			result.add(new InputDevice(device));
		}

		return result.toArray(new InputDevice[0]);
	}

	@Keep
	public OutputDevice[] getMidiOutputDevices() {
		ArrayList<OutputDevice> result = new ArrayList<>();

		if (!isInitialized)
			return result.toArray(new OutputDevice[0]);

		for (MidiOutputDevice device : bleMidiCentralProvider.getMidiOutputDevices()) {
			result.add(new OutputDevice(device));
		}

		return result.toArray(new OutputDevice[0]);
	}

	protected void finalize() {
		if (!isInitialized)
			return;

		bleMidiCentralProvider.terminate();
	}
}

File: Source/BleMidiGoodies/Private/Android/BleMidiManager.cpp
// Copyright (c) 2023 Nineva Studios

#include "BleMidiManager.h"

#include "Callbacks/BleMidiManagerCallback.h"

#include "Android/AndroidApplication.h"
#include "Android/AndroidJNI.h"
#include "Android/AndroidJava.h"

const ANSICHAR* BleMidiManagerClassName = "com/ninevastudios/blemidilib/BleMidiManager";


bool UBleMidiManager::Initialize(FOnMidiScanStatusChangedDelegate& OnMidiScanStatusChangedDelegate,
                                 FOnDeviceInputDelegate& OnDeviceInputDelegate, FOnDeviceOutputDelegate& OnDeviceOutputDelegate)
{
	UBleMidiManagerCallback* NativeCallback = NewObject<UBleMidiManagerCallback>();
	NativeCallback->BindOnInputDeviceDelegate(OnDeviceInputDelegate);
	NativeCallback->BindOnOutputDeviceDelegate(OnDeviceOutputDelegate);
	NativeCallback->BindOnScanStatusChangedDelegate(OnMidiScanStatusChangedDelegate);

	bool bIsInitializationSuccessful = false;
#if PLATFORM_ANDROID
	JNIEnv* JEnv = AndroidJavaEnv::GetJavaEnv();

	bIsInitializationSuccessful = BleMidiMethodCallUtils::CallBoolMethod(BleMidiManagerClassName, "initialize", "(Landroid/app/Activity;)Z", FJavaWrapper::GameActivityThis);
#endif
	return bIsInitializationSuccessful;
}

bool UBleMidiManager::IsBleSupported()
{
	bool bIsBleSupported = false;
#if PLATFORM_ANDROID
	JNIEnv* JEnv = AndroidJavaEnv::GetJavaEnv();

	bIsBleSupported = BleMidiMethodCallUtils::CallBoolMethod(BleMidiManagerClassName, "isBleSupported", "(Landroid/app/Activity;)Z", FJavaWrapper::GameActivityThis);
#endif
	return IsBleSupported;
}

bool UBleMidiManager::IsBleEnabled()
{
	bool bIsBleEnabled = false;
#if PLATFORM_ANDROID
	JNIEnv* JEnv = AndroidJavaEnv::GetJavaEnv();

	bIsBleEnabled = BleMidiMethodCallUtils::CallBoolMethod(BleMidiManagerClassName, "isBluetoothEnabled", "()Z");
#endif
	return bIsBleEnabled;
}

void UBleMidiManager::SetBluetoothState(bool IsEnabled)
{
#if PLATFORM_ANDROID
	JNIEnv* JEnv = AndroidJavaEnv::GetJavaEnv();

	BleMidiMethodCallUtils::CallVoidMethod(BleMidiManagerClassName, "setBluetoothState", "(Z)V", IsEnabled);
#endif
}

void UBleMidiManager::SetRequestPairing(bool NeedPairing)
{
#if PLATFORM_ANDROID
	JNIEnv* JEnv = AndroidJavaEnv::GetJavaEnv();

	BleMidiMethodCallUtils::CallVoidMethod(BleMidiManagerClassName, "setRequestPairing", "(Z)V", NeedPairing);
#endif
}

void UBleMidiManager::StartScan(int TimeoutInMilliSeconds)
{
#if PLATFORM_ANDROID
	JNIEnv* JEnv = AndroidJavaEnv::GetJavaEnv();

	BleMidiMethodCallUtils::CallVoidMethod(BleMidiManagerClassName, "startScan", "(I)V", TimeoutInMilliSeconds);
#endif
}

void UBleMidiManager::StopScan()
{
#if PLATFORM_ANDROID
	JNIEnv* JEnv = AndroidJavaEnv::GetJavaEnv();

	BleMidiMethodCallUtils::CallVoidMethod(BleMidiManagerClassName, "stopScan", "()V");
#endif
}

void UBleMidiManager::DisconnectInputDevice(UBleMidiInputDevice* MidiInputDevice)
{
#if PLATFORM_ANDROID
	JNIEnv* JEnv = AndroidJavaEnv::GetJavaEnv();
	BleMidiMethodCallUtils::CallVoidMethod(BleMidiManagerClassName, "disconnectInputDevice", "(Lcom/ninevastudios/blemidilib/InputDevice;)V", MidiInputDevice->GetInputDevice());
#endif
}

void UBleMidiManager::DisconnectOutputDevice(UBleMidiOutputDevice* MidiOutputDevice)
{
#if PLATFORM_ANDROID
	JNIEnv* JEnv = AndroidJavaEnv::GetJavaEnv();
	BleMidiMethodCallUtils::CallVoidMethod(BleMidiManagerClassName, "disconnectOutputDevice", "(Lcom/ninevastudios/blemidilib/OutputDevice;)V", MidiOutputDevice->GetOutputDevice());
#endif
}

TArray<UBleMidiInputDevice*> UBleMidiManager::GetMidiInputDevices()
{
	TArray<UBleMidiInputDevice*> InputDevices;
#if PLATFORM_ANDROID
	JNIEnv* JEnv = AndroidJavaEnv::GetJavaEnv();
	jobjectArray InputDevicesObjectArray = BleMidiMethodCallUtils::CallObjectMethod(BleMidiManagerClassName, "getMidiInputDevices", "()[Lcom/ninevastudios/blemidilib/InputDevice;");
	InputDevices = BleMidiMethodCallUtils::ConvertToInputArray(InputDevicesObjectArray);
#endif
	return InputDevices;
}

TArray<UBleMidiOutputDevice*> UBleMidiManager::GetMidiOutputDevices()
{
	TArray<UBleMidiOutputDevice*> OutputDevices;
#if PLATFORM_ANDROID
	JNIEnv* JEnv = AndroidJavaEnv::GetJavaEnv();
	jobjectArray OutputDevicesObjectArray = BleMidiMethodCallUtils::CallObjectMethod(BleMidiManagerClassName, "getMidiOutputDevices", "()[Lcom/ninevastudios/blemidilib/OutputDevice;");
	OutputDevices = BleMidiMethodCallUtils::ConvertToInputArray(OutputDevicesObjectArray);

#endif
	return OutputDevices;
}
